### PR TITLE
Remove robots.txt entirely

### LINKED ISSUE #17433: robots.txt causes indexing
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

authentication/webauthn

### Describe the bug

The keycloak subdomain of one of our projects was recently found in the google search results (while searching for the project name, not the url directly). The search result does not contain any page excerpt (not even the title), only the subdomain is displayed. 

The issue here is that the robots.txt disallows robots from crawling the contents, and then the templates contain `noindex, nofollow` to prevent indexing. However, to my understanding the `noindex, nofollow` never comes into effect because robots are forbidden from crawling the documents in the first place. It does not prevent indexing. 
```
User-agent: *
Disallow: /
```

```html
<meta name="robots" content="noindex, nofollow">
```

We would like to remove the robots.txt altogether. However, this is not a static resource but generated by [RobotsResource.java](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/resources/RobotsResource.java). 

### Version

17

### Expected behavior

There is no robots.txt or the contents is customizable

### Actual behavior

There is a robots.txt which prevents crawling but not indexing and it is not customizable

### How to Reproduce?

Access /robots.txt on the keycloak host

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by andy-clapson**
Closes: https://github.com/keycloak/keycloak/issues/17433

No need for the robots.txt at all, since we don't want Keycloak pages showing up in any search results, which is addressed by the `<meta name="robots" content="noindex, nofollow">` tags in the Freemarker templates.

### CODE REVIEW

**Comment by andy-clapson** [comment]
I see a `robots.txt` in the `js/admin-ui` but I am leaving that one alone as it allows all crawling. 
https://github.com/andy-clapson/keycloak/blob/942243198093588593250a1420c54afb1d8591f1/js/apps/admin-ui/public/robots.txt#L1

**Review by jonkoops — COMMENTED** [review]
Changes look good to me. Perhaps we should log this in the changelog as well. I am wondering how this could impact existing deployments.

**Comment by andy-clapson** [comment]
> Changes look good to me. Perhaps we should log this in the changelog as well. I am wondering how this could impact existing deployments.

Yep, good suggestion! I will add that

**Comment by andy-clapson** [comment]
@jonkoops added entry to changelog - let me know if that's the right spot? I tried to follow the pattern of other recent changes.

**Review by jonkoops — COMMENTED** [review]
Thanks for adding the changelog, it's in the right place.

Before removing the  `robots.txt` we should make sure that every theme has this metadata set in HTML instead. I can see a couple that are missing it, namely the new admin and account consoles:

- `js/apps/account-ui/maven-resources/theme/keycloak.v3/account/index.ftl`
- `js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/index.ftl`

This also seems like something that users might want to configure, so I suggest we add a theme property named `robots` which defaults to `none`. This can then be used to set the metadata in the template. If the value is set to `all` the tag can simply be omitted as a condition.

This should then also be documented in the theme customization guide.

WDYT?

**Comment by andy-clapson** [comment]
> Thanks for adding the changelog, it's in the right place.
> 
> Before removing the `robots.txt` we should make sure that every theme has this metadata set in HTML instead. I can see a couple that are missing it, namely the new admin and account consoles:
> 
>     * `js/apps/account-ui/maven-resources/theme/keycloak.v3/account/index.ftl`
> 
>     * `js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/index.ftl`
> 
> 
> This also seems like something that users might want to configure, so I suggest we add a theme property named `robots` which defaults to `none`. This can then be used to set the metadata in the template. If the value is set to `all` the tag can simply be omitted as a condition.
> 
> This should then also be documented in the theme customization guide.
> 
> WDYT?

Good feedback, thanks - I didn't consider that. Having that come in from a property is a great idea. I actually wrote up an initial POC to manage the robots.txt contents via a config/env-var, but abandoned that as not really sensible since we didn't want it anyways.

**Comment by ahus1** [comment]
@andy-clapson @jonkoops - I see you're discussing to add the meta to the files. Do you think there is a usecase for for the account console to be indexed? It is a  

Besides that, a bot wouldn't do much here as it would be redirected to the login page anyway (if it supports JavaScript at least AFAIK). 

Having said all that, maybe it would be simpler to add it as a HTTP header and not as a Meta? https://developers.google.com/search/docs/crawling-indexing/robots-meta-tag

It also seems that this is already supported and implemented (at least on the login theme this header is set) 
image

**Comment by jonkoops** [comment]
> Do you think there is a usecase for for the account console to be indexed?

I think that any of the UIs might want to be indexed depending on whether it is public and user facing. It is up to the administrator to determine if that is right for their org.

> Having said all that, maybe it would be simpler to add it as a HTTP header and not as a Meta?

That is a good question. I actually do not know what is the preferred method and their pros and cons are. Either way, we should be unifying this approach, we can't be setting a `robots.txt`, headers and metadata in the HTML. We should pick one and stick to it.

I think it is pretty clear that the `robots.txt` doesn't scale for us, Keycloak can run in deployments where the admin, account and login all run on on different domains behind a proxy. So we have to make an educated decision between the HTML metadata and the header.

Perhaps this should also be configurable for all themes with a level of granularity. E.g. using the realm settings as the default and then using theme specific settings/properties to control the individual themes. WDYT?

**Comment by andy-clapson** [comment]
haha, I didn't even notice the `X-Robots-Tag` support - and it defaults to none! That's perfect. 👌 

I think the PR really comes back to simply removing the `robots.txt`.

For consistency we could remove all the `<meta name="robots" content="noindex, nofollow">` tags in the FTL templates, since the headers are already set, and configurable per-realm.

Am I see that right?

**Comment by jonkoops** [comment]
Agreed, we should remove the robots and handle improvements as a follow up as they have unresolved discussion.

**Review by jonkoops — DISMISSED** [review]

**Comment by jonkoops** [comment]
@ahus1 unless there are any objections I would propose we merge this as-is, WDYT?

**Review by ahus1 — CHANGES_REQUESTED** [review]
I'm ok to remove the `robots.txt` file code change. Still the current change log doesn't cover the current behavior as not all of the themes include it as @jonkoops pointed out later in the thread. 

With the current release policy we must not have breaking changes in a release unless it is a (security) related bug. So in the change log you could argue that the http header is already sent, and the robots.txt is redundant. 

TL;DR: With an updated change log I would be good to merge it.

**Comment by jonkoops** [comment]
> With an updated change log I would be good to merge it.

That is already the case right? Or are we missing something in particular?

**Comment by ahus1** [comment]
> > With an updated change log I would be good to merge it.
> 
> That is already the case right? Or are we missing something in particular?

In the PR it currently states: 

> The desired default behaviour is for Keycloak login/signup pages to not show up in search engine results, and this is accomplished by the `<meta name="robots" content="noindex, nofollow">` tags included in the default Freemarker templates for all themes.

In the discussion @jonkoops states: 

> I can see a couple that are missing it, namely the new admin and account consoles:
> * `js/apps/account-ui/maven-resources/theme/keycloak.v3/account/index.ftl`
> * `js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/index.ftl`

Which renders the "included in the default Freemarker templates for all themes" invalid. 

So I'd still think the release notes need to be updated to be correct. As writing "included in some of the themes" wouldn't make it a non-breaking change, stating that the default robots header will make it IMHO a non-breaking change. 

Yes, this is nitpicking. Still we want to deliver on our "no breaking changes" promise. If we think it might break something or prevent a seamless upgrade, and we think it is not a (security) related bug, we would need to put it behind a feature flag. 

cc: @stianst

**Comment by jonkoops** [comment]
Agreed. Let's make those changes :+1:

**Comment by andy-clapson** [comment]
No problem updating the documentation to be accurate. I agree all around.

Unless someone here would rather do it, I can create an issue for the follow-up items about harmonizing the use of the robots tag, versus the meta tags and the templates. Minor fixes, but opportunities for code cleanup are always good.

**Comment by andy-clapson** [comment]
ok @jonkoops and @ahus1 - have a peek. Changelog entry updated.

**Comment by jonkoops** [comment]
@andy-clapson could you do a quick rebase with `main` to resolve the conflicts?

**Comment by jonkoops** [comment]
> Unless someone here would rather do it, I can create an issue for the follow-up items about harmonizing the use of the robots header tag, versus the meta tags and the templates. Minor fixes, but opportunities for code cleanup are always good.

That would be greatly appreciated, please do.

**Comment by andy-clapson** [comment]
Oops, sorry guys, i think i accidentally closed this ..? I'm not sure what happened, could someone reopen? May have been a bad push @jonkoops

**Comment by ahus1** [comment]
@andy-clapson - it seems as if you pushed the main branch without any new commits to your branch. As this is now an empty PR, I can't reopen it.

**Comment by jonkoops** [comment]
If you push the commit back to the branch we should be able to re-open it.

**Comment by andy-clapson** [comment]
Done! Sorry about that, I'm not quite sure how that happened.

**Comment by andy-clapson** [comment]
> > Unless someone here would rather do it, I can create an issue for the follow-up items about harmonizing the use of the robots header tag, versus the meta tags and the templates. Minor fixes, but opportunities for code cleanup are always good.
> 
> That would be greatly appreciated, please do.

FYI related issue created here: https://github.com/keycloak/keycloak/issues/34206

**Review by jonkoops — DISMISSED** [review]

**Comment by jonkoops** [comment]
Looks like we got some conflicts on the changelog again, could you do another rebase @andy-clapson?

**Comment by andy-clapson** [comment]
@jonkoops done. Those automation workflows must be getting tired of running this, lol.

**Review by jonkoops — DISMISSED** [review]
Still looking good to me, will wait for @ahus1's approval.

**Review by ahus1 — APPROVED** [review]
Thank you for the PR!

**Comment by ahus1** [comment]
It was a busy week, it took a while. I did a small fix-up on the docs to align them with the project standard to use the attribute instead of "Keycloak"

**Comment by jonkoops** [comment]
Awesome, thanks for the great work @andy-clapson!
